devicetypes/heltun/heltun-tps05-switch.src/heltun-tps05-switch.groovy

```diff
@@ -0,0 +1,680 @@
+		def child = childDevices?.find {channelNumber(it.deviceNetworkId) == buttonId }
+		child?.sendEvent([name: "button", value: state, data: [buttonNumber: 1], isStateChange: true])
+	}
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.multichannelv3.MultiChannelCmdEncap cmd) {
+	def numberOfButtons = state.numberOfButtons
+	def endPoint = cmd.sourceEndPoint.toInteger()
+	def encapsulatedCommand = cmd.encapsulatedCommand([0x32: 3, 0x25: 1, 0x20: 1])
+	def value = encapsulatedCommand.value
+	def childDevice = childDevices?.find {channelNumber(it.deviceNetworkId) == endPoint }
+	def corrRelCons = 0
+	def corRelParam = 11 + endPoint - numberOfButtons
+	if (endPoint in numberOfButtons..(2*numberOfButtons)) {
+		def param = parameterMap().find( {it.paramNum == corRelParam } ).name
+		def paramState = state."$param"
+		if (paramState) {
+			corrRelCons = paramState.value
+		}
+	}
+	if (childDevice) {
+		childDevice.sendEvent(name: "switch", value: value ? "on" : "off")
+		if (value) {
+			sendEvent(name: "switch", value: "on")
+			childDevice.sendEvent(name: "power", value: corrRelCons, unit: "W")
+		} else {
+			childDevice.sendEvent(name: "power", value: 0, unit: "W")
+			if (!childDevices.any { it.currentValue("switch") == "on" }) {
+				sendEvent(name: "switch", value: "off")
+			}
+		}
+	}
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.clockv1.ClockReport cmd) {
+	def currDate = Calendar.getInstance(location.timeZone)
+	def time = [hour: currDate.get(Calendar.HOUR_OF_DAY), minute: currDate.get(Calendar.MINUTE), weekday: currDate.get(Calendar.DAY_OF_WEEK)]
+	if ((time.hour != cmd.hour) || (time.minute != cmd.minute) || (time.weekday != cmd.weekday)) {
+		sendHubCommand(zwave.clockV1.clockSet(time).format())
+	}
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.multichannelassociationv2.MultiChannelAssociationReport cmd) {
+	def cmds = []
+	if (cmd.groupingIdentifier == 1) {
+		if (cmd.nodeId != [0, zwaveHubNodeId, 0]) {
+			cmds << zwave.multiChannelAssociationV2.multiChannelAssociationRemove(groupingIdentifier: 1).format()
+			cmds << zwave.multiChannelAssociationV2.multiChannelAssociationSet(groupingIdentifier: 1, nodeId: [0,zwaveHubNodeId,0]).format()
+		}
+	}
+	if (cmds) {
+		sendHubCommand(cmds, 1200)
+	}
+}
+
+def configure() {
+	refresh()
+}
+
+def refresh() {
+	def numberOfButtons = state.numberOfButtons
+	def cmds = []
+	cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType:1).format() //roomTemperature
+	cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType:3).format() //Humidity
+	cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType:5).format() //Illuminance
+	for (i in 1..(2 * numberOfButtons)) {
+		cmds << encap(zwave.switchBinaryV1.switchBinaryGet(), i)
+	}
+	cmds << zwave.clockV1.clockGet().format()
+	cmds << zwave.multiChannelAssociationV2.multiChannelAssociationGet(groupingIdentifier: 1).format()
+	sendHubCommand(cmds, 1200)
+	runIn(15, "checkParam")
+}
+
+def ping() {
+	refresh()
+}
```

Comment:
I don't think if you need whole `refresh()` here, especially when that includes checking configuration parameters.
Remember that `ping()`'s responsibility is mainly to maintain device's `online`/`offline` status.